cluster-autoscaler/estimator/pod_priority_processor.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package estimator
+
+import (
+	"sort"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+)
+
+// PodScoreInfo contains Pod and score that corresponds to how important it is to handle the pod first.
+type PodScoreInfo struct {
+	score float64
+	pod   *apiv1.Pod
+}
+
+// PodPriorityProcessor is an interface used to determine the order of the pods
+// used while binpacking during scale up estimation
+type PodPriorityProcessor interface {
+	Process(pods []*apiv1.Pod, nodeTemplate *framework.NodeInfo) []*apiv1.Pod
+}
+
+// DefaultPodPriorityProcessor is the default implementation of the pod priority processor
+// It refactors the existing implementation of sorting pods by pod score
+type DefaultPodPriorityProcessor struct {
+}
+
+// NewDefaultPodPriorityProcessor returns the object of DefaultPodPriorityProcessor
+func NewDefaultPodPriorityProcessor() *DefaultPodPriorityProcessor {
+	return &DefaultPodPriorityProcessor{}
+}
+
+// Process is the processing func that sorts the pods based on the size of the pod
+func (d *DefaultPodPriorityProcessor) Process(pods []*apiv1.Pod, nodeTemplate *framework.NodeInfo) []*apiv1.Pod {
+	podInfos := make([]*PodScoreInfo, 0, len(pods))
+	for _, pod := range pods {
+		podInfos = append(podInfos, d.CalculatePodScore(pod, nodeTemplate))
+	}
+	sort.Slice(podInfos, func(i, j int) bool { return podInfos[i].score > podInfos[j].score })
+	podList := make([]*apiv1.Pod, 0, len(pods))
+	for _, podInfo := range podInfos {
+		podList = append(podList, podInfo.pod)
+	}
+
+	return podList
+}
+
+// CalculatePodScore score for  pod and returns PodScoreInfo structure.
+// Score is defined as cpu_sum/node_capacity + mem_sum/node_capacity.
+// Pods that have bigger requirements should be processed first, thus have higher scores.
+func (d *DefaultPodPriorityProcessor) CalculatePodScore(pod *apiv1.Pod, nodeTemplate *framework.NodeInfo) *PodScoreInfo {
```

Comment:
Any reason for:
1. Exporting this method? Do we plan on using it outside of this package?
2. Having this as a method and not just a function? It doesn't use `d` in any way.

Reply:
> Having this as a method and not just a function? It doesn't use d in any way.
Yes. This one I specifically added as a method and not just a function. The idea behind it is, the method of calculation is "core functionality" of the object and rather not a utility for the processor, which is why I thought this should be attached to the object even though it doesn't use `d`. I'm okay to detach it.

> Exporting this method? Do we plan on using it outside of this package?
I think this was similar reason as `PodScoreInfo` being exported for possible extension of this implementation. However, I don't plan on using it at this moment. Will make the internal.

Reply:
I don't think that's necessary, you're explanation makes sense, thanks.